KIPs/kip-dreaft_sdk_common_architecture.md

```diff
@@ -0,0 +1,1623 @@
+
+#### KeyringContainer
+
+`KeyringContainer` is a class that manages SingleKeyring, MultipleKeyring, and RoleBasedKeyring instances.
+
+##### Variable description
+
+| Variable | Description |
+| ----------- | ----------- |
+| length: int | The number of keyrings in keyringContainer. |
+| addressKeyringMap: Map<String, AbstractKeyring> | A Map that has an account address as a key and a Keyring instance corresponding to that address as a value. |
+
+##### Method description
+
+| Method | Description |
+| ----------- | ----------- |
+| generate(num: int, entropy: String): List&#60;String&#62; | Generates instances of SingleKeyring in the keyringContainer with randomly generated private keys. |
+| add(keyring: AbstractKeyring): AbstractKeyring | Adds an instance of keyring to the keyringContainer. It throws an exception if the newly given keyring has the same address with one of the keyrings that already exist in `caver.wallet`. |
+| newKeyring(address: String, privateKeyString: String): AbstractKeyring | Creates a keyring instance with given parameters and adds it to the `caver.wallet`. If key is a private key string, a SingleKeyring instance that uses a single private key is created. The keyring created is added to `caver.wallet`. It throws an exception if the address string or private key string is invalid. |
+| newKeyring(address: String, privateKeyArray: List&#60;String&#62;): AbstractKeyring | Creates a keyring instance with given parameters and adds it to the `caver.wallet`. If key is an array containing private key strings, a MultipleKeyring instance that use multiple private keys is created. The keyring created is added to `caver.wallet`. It throws an exception if the address string or private key strings are invalid. |
+| newKeyring(address: String, roleBasedPrivateKeyArray: List&#60;List&#60;String&#62;&#62;): AbstractKeyring | Creates a keyring instance with given parameters and adds it to the `caver.wallet`. If key is a 2D array of which each element contains the private key(s) to be used for each role, a RoleBasedKeyring instance is created. The keyring created is added to `caver.wallet`. It throws an exception if the address string or private key strings are invalid. |
+| updateKeyring(keyring: AbstractKeyring): AbstractKeyring | Updates the keyring inside the keyringContainer. It throws an exception if the matching keyring is not found. |
+| getKeyring(address: String): AbstractKeyring | Returns the keyring instance corresponding to the address. |
+| sign(address: String, transaction: AbstractTransaction): AbstractTransaction | Signs the transaction as a sender of the transaction and appends signatures in the transaction object using the keyring in keyringContainer. This method will use all the private keys. It throws an exception if the keyring to be used for signing cannot be found in `caver.wallet`. |
```

Comment:
```suggestion
| sign(address: String, transaction: AbstractTransaction): AbstractTransaction | Signs the transaction as a sender of the transaction and appends signatures in the transaction object using the keyring associated with the given address in keyringContainer. This method will use all the private keys in the keyring. It throws an exception if the keyring to be used for signing cannot be found in the `KeyringContainer` instance. |
```

Reply:
@jimni1222 What happens if the address and the transaction sender are different?

Reply:
An exception will be thrown.

Reply:
@jimni1222 Then, please add one more sentence.
```
It also throws an exception if `address` is different from the sender address of `transaction`.
```

Reply:
I modified the last sentence :)